### ISSUE TITLE
Introduce `MatchLabelKeys` (the same one as Pod Topology Spread) to `PodAffinityTerm`

### ISSUE THREAD

**Issue by sanposhiho** [kubernetes/kubernetes#112947]
### What would you like to be added?

Introduce `MatchLabelKeys` to [PodAffinityTerm](https://pkg.go.dev/k8s.io/api/core/v1#PodAffinityTerm) (like [we do to Pod Topology Spread](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades)).

### Why is this needed?

Introduce `MatchLabelKeys` (like [we introduce it to Pod Topology Spread](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades)) to [PodAffinityTerm](https://pkg.go.dev/k8s.io/api/core/v1#PodAffinityTerm).

---

Currently, the scheduler doesn’t treat `Terminating` Pods as removed Pods.
See https://github.com/kubernetes/kubernetes/issues/105661. Pod Topology Spreading gets imbalanced after Deployment/Replicaset upgrades because of that. So, we introduced `MatchLabelKeys` to Pod Topology Spread (https://github.com/kubernetes/enhancements/issues/3243).

In Pod Affinity and Pod Anti Affinity, it gives a negative impact on the scheduling result to take old Pods into account during scheduling (“Old Pods” = Pods from a old deployment); Old pods may result in filtering out Nodes that should have gone to the Scoring phase (requiredDuringSchedulingIgnoredDuringExecution), or giving low scores to Nodes that should have been given higher scores. (preferredDuringSchedulingIgnoredDuringExecution).

/sig scheduling
/kind feature
/assign

**Comment by sanposhiho** [comment]
cc @kubernetes/sig-scheduling-leads

**Comment by alculquicondor** [comment]
+1 it's good to have API parity

**Comment by kerthcet** [comment]
We discussed a bit about this before https://kubernetes.slack.com/archives/C09TP78DV/p1661093052989079?thread_ts=1661006326.904089&cid=C09TP78DV, and I was thinking why don't we unify the label names in deployment and statefulset(currently they are `pod-template-name` and `controller-revision-hash`), then no need for this matchLabelKeys anymore. I think this is the main case why we need matchLabelKeys for identifying old/new pods.

For third-party controllers, they can also follow the rule.

**Comment by kerthcet** [comment]
The key point is kubernetes should provide a mechanism to make scheduler aware of the old/new replicas. Currently, the hash label seems can provide some help.

**Comment by sanposhiho** [comment]
Found an old issue and seems it has been implemented to use different labels as intended.

> We can use controller-revision-hash for DaemonSet and StatefulSet, and keep Deployment as-is. Deployment hash label key should be different from other two because Deployment pods hash labels is used by selector (unlike the other two controllers).
https://github.com/kubernetes/kubernetes/issues/47554

Adding a new well-known label to identify Pod's group might be one general solution instead of unifying the labels of deployment and statefulset. Actually, it's been also discussed when we introduced `MatchLabelKeys`.  (option 1 in the doc.)
https://docs.google.com/document/d/12F1YXsvy4SmumPiL3LHMLnuMHvyMqSzDqF-L9J37ZfA/edit#

**Comment by alculquicondor** [comment]
Right, because changing the default behavior is not backwards compatible. I think `MatchLabelKeys` gives more flexibility to users. However, it would still be useful to have a well known label, even if they are not the same for each workload API.

**Comment by Huang-Wei** [comment]
> I was thinking why don't we unify the label names in deployment and statefulset(currently they are pod-template-name and controller-revision-hash)

Implicit unifying may sometimes misinterpret user workloads' intentions. For example, if user's intention is to spread app1 AND app2 (both apps with identical labelSelector). During app2's rolling upgrade, with `MatchLabelKeys`, it's user's responsibility to explicitly specify whether to spread app1 + app2's new Pods, or just app2's new Pods, or whatever intention the user wants. So I'm more in favor of an explicit API field.

**Comment by ahg-g** [comment]
Adding MatchLabelKeys to other scheduling primitives is a reasonable next step.

Regarding using a common label: we briefly discussed that in the KEP as an alternative. Labels are arbitrary, the scheduler shouldn't really make such strong assumptions about their semantics. What is the concern about MatchLabelKeys?

**Comment by kerthcet** [comment]
Thanks for feedbacks, unifying the labels may not be a good choice. 

About `MatchLabelKeys`, I have several concerns:
1) Currently we treat it as a feature option, but I can't image that people will expect scheduling decisions taking read-to-rolling-update pods into consideration. We could say it's not for breaking end-users, but actually the impact is small for that's what they want. 

2) As @ahg-g said, we'd like to apply the `MatchLabelKeys` to other scheduling primitives, it grows the burden of users for they have to configure this for all corresponding plugins(And I think most of the plugins yes). 

3) Besides identifying the old/new replicas in rolling update, does `MatchLabelKeys` have other earnings, from the https://github.com/kubernetes/enhancements/issues/3243., I got no other answers. Please add other cases if you guys have.

**Comment by alculquicondor** [comment]
In general, MatchLabelKeys can be used to simplify your affinity selectors. Now you only have to specify the label key, instead of having to copy the label value as well.

So that should be some motivation for users to migrate to the new API.

**Comment by moustafab** [comment]
I think the vast majority of pod anti affinity users use it to spread their workloads for availability reasons (across nodes or AZs) so `MatchLabelKeys` is a much easier way to specify that, and much more resilient to changes in naming for example for a workload. A single generic antiAffinity term can be applied for a given labeling convention and that's much more powerful and less repetitive.

**Comment by sanposhiho** [comment]
Looks worth making a concrete proposal for it, as there are more positive thoughts than negative ones so far.
Opened https://github.com/kubernetes/enhancements/issues/3633, we can continue to discuss it in KEP.

**Comment by sanposhiho** [comment]
/remove-lifecycle stale

It's on going
